# Week 4: HealthConnect 
## Machine Learning problem definition

## 1. Project and problem understanding

### 1.1 Project context

HealthConnect Clinic is a fictional outpatient healthcare provider that manages appointment-based services. The clinic is experiencing missed appointments and wants to use data and AI to improve appointment attendance and patient support.

As part of the Data Science track, this project will assess whether historical appointment data can support a machine learning solution for predicting appointment no-shows.

### 1.2 Role within the HealthConnect project

As a Data Science intern, my role is to define and assess a machine learning problem that can support HealthConnect Clinic's broader objective of reducing missed appointments and improving patient support. The focus of my contribution is to determine whether the available appointment data can support a no-show prediction solution and to establish an initial modelling approach for subsequent development.

### 1.3 Resources used

The following resources were reviewed and used during the Week 4 assessment:

- HealthConnect appointment dataset (HealthConnect_Appointment_Data.csv)
- HealthConnect data dictionary (HealthConnect_Data_Dictionary.xlsx)
- Week 4 HealthConnect Experience Lab assignment

The appointment dataset was used for the initial data assessment, while the Data Dictionary was used to understand the variables and identify relevant potential features. The Week 4 assignment was used to guide the problem definition, proposed target, feature selection, modelling approach, and key considerations.

### 1.4 Data Science problem

The specific Data Science problem is to determine whether historical HealthConnect appointment data can be used to predict whether a scheduled appointment will result in a no-show.

This is a supervised machine learning problem because the historical appointment records contain information about appointments and their observed outcomes. The proposed solution will focus on classification, with the aim of distinguishing appointments that are likely to be attended from those that are likely to be missed.

### 1.5 Project objective

To assess the suitability of the available appointment data for developing a no-show prediction model. The analysis will identify the proposed prediction target, relevant input features, and an appropriate initial modelling approach that can support HealthConnect in understanding and identifying appointments at risk of being missed.

## 2. Initial data assessment

### 2.1 Dataset overview

In [5]:
import pandas as pd

In [8]:

df = pd.read_csv("HealthConnect_Appointment_Data.csv")

In [9]:
df.shape

(5000, 18)

The HealthConnect appointment dataset contains 5,000 records and 18 variables. Each record represents an appointment and includes information relating to patient characteristics, appointment details, booking information, previous appointment history, reminders, distance to the clinic, waiting time, and appointment outcome.

In [10]:
df.columns   #checking column names

Index(['appointment_id', 'patient_id', 'gender', 'age', 'age_group',
       'appointment_type', 'booking_date', 'appointment_date',
       'appointment_day', 'appointment_time', 'booking_lead_days',
       'previous_appointments', 'previous_no_shows', 'reminder_sent',
       'reminder_channel', 'distance_to_clinic_km', 'waiting_time_minutes',
       'appointment_outcome'],
      dtype='str')

### 2.2 Dataset structure

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   appointment_id         5000 non-null   str    
 1   patient_id             5000 non-null   str    
 2   gender                 5000 non-null   str    
 3   age                    5000 non-null   int64  
 4   age_group              5000 non-null   str    
 5   appointment_type       5000 non-null   str    
 6   booking_date           5000 non-null   str    
 7   appointment_date       5000 non-null   str    
 8   appointment_day        5000 non-null   str    
 9   appointment_time       5000 non-null   str    
 10  booking_lead_days      5000 non-null   int64  
 11  previous_appointments  5000 non-null   int64  
 12  previous_no_shows      5000 non-null   int64  
 13  reminder_sent          5000 non-null   str    
 14  reminder_channel       3634 non-null   str    
 15  distance_to_cli

### 2.3 Data quality assessment

In [13]:
df.isnull().sum()   #missing values

appointment_id              0
patient_id                  0
gender                      0
age                         0
age_group                   0
appointment_type            0
booking_date                0
appointment_date            0
appointment_day             0
appointment_time            0
booking_lead_days           0
previous_appointments       0
previous_no_shows           0
reminder_sent               0
reminder_channel         1366
distance_to_clinic_km      90
waiting_time_minutes       60
appointment_outcome         0
dtype: int64

In [14]:
df.duplicated().sum()   #duplicates

np.int64(0)

In [15]:
df["appointment_outcome"].value_counts()   #target values check

appointment_outcome
No-Show      2423
Attended     2314
Cancelled     263
Name: count, dtype: int64

In [16]:
df["appointment_outcome"].value_counts(normalize= True)*100   #percentage distribution of these outcomes

appointment_outcome
No-Show      48.46
Attended     46.28
Cancelled     5.26
Name: proportion, dtype: float64

In [17]:
df["appointment_outcome"].unique()   #whether the appointment outcome contains unexpected values

<ArrowStringArray>
['No-Show', 'Attended', 'Cancelled']
Length: 3, dtype: str

#### Data quality findings

The initial data quality assessment identified 1,366 missing values in reminder_channel, 90 missing values in distance_to_clinic_km, and 60 missing values in waiting_time_minutes. No completely duplicated records were identified. The appointment_outcome variable contains three valid categories: No-Show, Attended, and Cancelled. No unexpected outcome categories were observed.

### 2.4 Data suitability

The dataset appears potentially suitable for investigating appointment no-show prediction because it contains 5,000 appointment records, a defined appointment outcome, and several variables that may provide relevant information about patient characteristics, appointment details, booking behaviour, previous appointment history, reminder information, distance to the clinic, and waiting time.

However, the suitability of the data for a reliable machine learning model cannot be fully established at this stage. Missing values, variable relationships, the treatment of cancelled appointments, and potential modelling issues will need to be considered during subsequent project stages.

## 3. Proposed target variable

### 3.1 Target variable

The proposed target variable is appointment_outcome. The variable records the final status of each appointment and contains three categories: Attended, No-Show, and Cancelled.

The primary outcome of interest is No-Show, because the objective is to investigate whether an appointment can be identified as being at risk of being missed.

### 3.2 Handling appointment outcomes

For the initial no-show prediction problem, the target will be treated as a binary classification outcome. Appointments recorded as No-Show will represent the positive class, while appointments recorded as Attended will represent the negative class.

Appointments recorded as Cancelled will be excluded from the initial prediction target because a cancellation represents a different appointment outcome from a patient failing to attend. The treatment of cancelled appointments can be revisited in a later stage if the project requires a separate cancellation prediction or a broader non-attendance analysis.

No records will be removed or target values transformed during Week 4 and this represents the proposed approach for subsequent modelling stages.

## 4. Potential features

### 4.1 Potential input features

Potential input features were identified based on their relevance to patient characteristics, appointment details, booking behaviour, previous appointment history, reminder information, and accessibility or operational factors. The initial candidate features are gender, age, age_group, appointment_type, appointment_day, appointment_time, booking_lead_days, previous_appointments, previous_no_shows, reminder_sent, reminder_channel, distance_to_clinic_km, and waiting_time_minutes.

Identifier variables such as appointment_id and patient_id are not proposed as predictive features because they primarily identify appointments or patients. The appointment_outcome variable is reserved as the target. The booking and appointment dates may be considered later for deriving useful time-related features, but no feature engineering is proposed at the Week 4 stage.

### 4.2 Feature considerations

Some potential features may require further assessment before they are used for modelling. In particular, age_group is derived from age, while booking_date and appointment_date may be used to derive additional time-related information. Variables such as waiting_time_minutes should also be assessed for potential information leakage because their availability at the time of prediction must be considered.

These considerations will be investigated during the subsequent modelling stage rather than being resolved during Week 4.

## 5. Initial modelling approach

### 5.1 Proposed modelling strategy

The proposed machine-learning approach is a supervised binary classification problem. The model will use historical appointment information to predict whether a scheduled appointment is likely to result in a No-Show or an Attended outcome.

The initial modelling stage will involve preparing the selected features, separating the data into training and testing sets, and comparing suitable classification algorithms. The final model selection will be based on appropriate evaluation metrics rather than accuracy alone.

### 5.2 Proposed data preparation

The selected features will require appropriate preprocessing before they can be used for machine-learning. This is expected to include handling missing values, converting categorical variables into a suitable numerical representation, and ensuring that date-related information is represented appropriately where required.

The proposed binary target will also need to be encoded into a numerical form during the modelling stage. Any preprocessing decisions will be applied using only information available from the appropriate training data to reduce the risk of data leakage.

### 5.3 Model evaluation

The model will be evaluated using classification metrics that provide insight into its ability to identify no-show appointments. These may include precision, recall, F1-score, and a confusion matrix, with particular attention given to recall for the No-Show class. Accuracy may also be reported but will not be used as the sole measure of model performance.

The final evaluation approach will be confirmed during the modelling stage after the class distribution and model performance have been assessed.

## 6. Key modelling considerations

### 6.1 Missing data

Missing values will need to be addressed before model development. The treatment should consider the meaning of the missing values rather than applying the same method to all variables. In particular, the missing values in reminder_channel require careful consideration because the Data Dictionary indicates that None may occur when no reminder was sent. The appropriate treatment of missing values will therefore be determined during the data preparation stage.

### 6.2 Class distribution

The appointment outcomes are distributed across No-Show (48.46%), Attended (46.28%), and Cancelled (5.26%). Since the proposed prediction problem focuses on No-Show versus Attended outcomes, cancelled appointments will be treated separately as described in Section 3.2. The resulting class distribution should be assessed again during model development to determine whether any class imbalance requires consideration.

### 6.3 Data leakage

Data leakage must be considered when selecting and preparing features. Variables should only be used if the information would be available at the time a no-show prediction is made. In particular, waiting_time_minutes requires further assessment because information about actual or estimated waiting time may not be available at the prediction point. The timing and availability of other variables will also be reviewed during model development to reduce the risk of leakage.

### 6.4 Limitations and risks

The dataset is synthetic and anonymised, so the findings may not fully represent real world patient behaviour or clinic operations. The dataset also contains missing values in selected variables, which will require appropriate treatment during model development. In addition, the treatment of cancelled appointments and the availability of certain variables at the prediction point may affect the final modelling approach. These limitations will be considered when interpreting model results and assessing the potential usefulness of the proposed solution.

## 7. Week 4 project summary

### 7.1 Key observations

The Week 4 assessment established that the HealthConnect appointment dataset contains 5,000 records and 18 variables and includes a defined appointment outcome. The data contains information that may be relevant to no-show prediction, including patient characteristics, appointment details, booking behaviour, previous appointment history, reminder information, distance to the clinic, and waiting time.

The assessment also identified missing values in three variables and confirmed that there are no completely duplicated records. The appointment outcome contains three categories: No-Show, Attended, and Cancelled. For the proposed prediction problem, No-Show and Attended will form the initial binary classification target, while Cancelled appointments will be treated separately.

### 7.2 Proposed next stage

The next stage will focus on preparing the dataset for modelling and developing the proposed no-show classification approach. This will include appropriate treatment of missing values, preparation of the selected features, handling of the target variable, and evaluation of suitable classification models using appropriate performance metrics.